Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,459 @@
+        // will be possible to use to bind the kernel poller thread to a given cpu (and share one amongst all rings)
+        // - not yet in the kernel and work in progress, but should be a better fit and worth trying.
+        // Alternatively we could look at limiting number of threads created to numcpu / 2 or so
+        // instead of starting one thread per core in the multithreaded event loop
+        // or allowing customization of whether to use SQPOLL somewhere higher up.
+        // Also, current IORING_SETUP_SQPOLL requires correct privileges to run (root or specific privs set)
+        if (CNIOLinux.CNIOLinux_io_uring_queue_init(ringEntries, &ring, 0 ) != 0) // IORING_SETUP_SQPOLL
+         {
+             throw UringError.uringSetupFailure
+         }
+        
+        _debugPrint("io_uring_queue_init \(self.ring.ring_fd)")
+     }
+  
+    @inline(never)
+    public func io_uring_queue_exit() {
+        _debugPrint("io_uring_queue_exit \(self.ring.ring_fd)")
+        CNIOLinux_io_uring_queue_exit(&ring)
+    }
+
+    //   Ok, this was a real bummer - turns out that flushing multiple SQE:s
+    //   can fail midflight and this will actually happen for real when e.g. a socket
+    //   has gone down and we are re-registering polls this means we will silently lose any
+    //   entries after the failed fd. Ouch. Proper approach is to use io_uring_sq_ready() in a loop.
+    //   See: https://github.com/axboe/liburing/issues/309
+            
+    //   FIXME: NB we can get stuck in this flush for synthetic test that does a massive
+    //   amount of registration modifications. Basically, this is due to back pressure from uring
+    //   to make us reap CQE:s as the CQE queue would be full with responses. To avoid this
+    //   we should either limit the amount of outbound operations we generate before reaping
+    //   CQE:S, or run with IORING_SETUP_SQPOLL (then we would not get stuck here, as the kernel
+    //   would simply stop reading SQE:s if the CQE queue is full, we would then instead get
+    //   stuck trying to get a new SQE, which also would be a problem.). So, better limit amount
+    //   of generated outbound operations sent and allow us to reap CQE:s somehow.
+    //   Current workaround is to set up a fairly big ring size to avoid getting stuck, but
+    //   that gives us worse locality of reference for the memory used by the ring, so worse
+    //   cache behavior. Should be revisited, probably a more reasonable size of the ring
+    //   would be in the hundreds rather than thousands.
+    
+    @inline(never)
+    public func io_uring_flush() {         // When using SQPOLL this is just a NOP
+        var submissions = 0
+        var retval : Int32
+        
+        _debugPrint("io_uring_flush")
+
+        // FIXME:  it may return -EAGAIN or -ENOMEM when there is not enough memory to do internal allocations.
```

Comment:
Are these kernel space allocation failures or userspace ones?

Reply:
They are kernel allocations, and there is also one more tricky case - uring consists of two ring buffers, one for submitted IO (SQE queue) and one for completions (CQE). If the application doesn't reap the CQE buffer, the kernel will use back pressure on the application by not accepting any additional SQE:s. This is what gives us EAGAIN here. For well-behaved apps, not a problem, but for e.g. the tests that flip registrations back/forth 1000 times, we generate tons of SQE:s before starting processing CQE:s again, so we can get EAGAIN.

Reply:
I've updated it so that we don't fail on EAGAIN/EBUSY here (it will give us another chance at reaping CQE:s and the SQE:s will be flushed at the next go around of whenReady(), we will not lose the SQE:s, they are just waiting for processing in the queue to the kernel). The memory allocation kernel side could still fail, I put in a 1s sleep between retries, but I'd be fine with a fatalError here too.